### ISSUE TITLE
Remove peculiar comment on Applicative (UTerm t) instance

### ISSUE THREAD

**Issue by treeowl** [wrengr/unification-fd#5]
There's a peculiar comment about the `Applicative (UTerm t)` instance maybe not making sense because the first argument to `(<*>)` will be a `UTerm` with a function in it. That weirdness largely goes away with `liftA2`; for GHC 8.2, you can add

```haskell
    liftA2 f (UVar a) (UVar b) = UVar (f a b)
    liftA2 f (UVar a) (UTerm mb) = UTerm (fmap (f a) <$> mb)
    liftA2 f (UTerm ma) b = UTerm ((\q -> liftA2 f q b) <$> ma)
```

I can't say if it's *useful* or not, but it seems pretty harmless. I'm not sure the `Alternative` instance is even harmless; the `free` package documents that one as not obeying the `Alternative` laws, though I don't know of any standard set thereof.

**Comment by wrengr** [comment]
iirc, I stole the `Alternative` instance (along with most the others) from the free package, mostly just so folks wouldn't complain about them being missing (as evidenced by all the ones with comments about them not necessarily making sense). It's been a few years since I've looked at the package, so I'll take a look and see if things seem more sensible now than they used to

**Comment by wrengr** [comment]
Re `Alternative` and `MonadPlus`, yeah, the version Ed has in `free` is really broken— it's not even a monoid!

I do think it can be fixed. Conceptually, we'd want to think of the underlying `Alternative t` as providing us a choice (or more generally, a "sum") of `t` structures. Thus, each individual term can actually be thought of as a superposition of terms packed together to share prefixes. With this idea in mind we can use the definition `UTerm a <|> UTerm b = UTerm (a <|> b)` which satisfies the monoid laws assuming the underlying instance does. The only problem is this doesn't handle 'UVar'... To do that in a fully general way we'd also require that `v` forms a monoid, which is also a two-sided action over `t`. This is doable,... but it'd mean we'd need to generalize the `Variable` class so that it can handle such choice/sum variables. (The `BindingMonad` class would also, in principle, need generalizing; but it's already general enough, since all the functions are monadic and we can choose an `m` to support the nondeterminism/summation of the variables.) Again, it's doable. But it'd require going through everything to make sure we don't violate this sort of semantics elsewhere. Would be a nice feature to add someday, perhaps after I finish my ABT library for case analysis.

tl;dr: removing the instances for now.
